src/crypto/tls/ech.go

```diff
@@ -327,29 +326,37 @@ func (c *Conn) echAcceptOrReject(hello *clientHelloMsg) (*clientHelloMsg, error)
 	return helloInner, nil
 }
 
-// echClient represents a ClientECH structure, the payload of the client's
-// "encrypted_client_hello" extension.
-type echClient struct {
+// echClientOuter represents a ClientECH structure, the payload of the client's
+// "encrypted_client_hello" extension that appears in the outer handshake.
+type echClientOuter struct {
 	raw []byte
 
 	// Parsed from raw
 	handle  echContextHandle
 	payload []byte
 }
 
-// echUnmarshalClient parses a ClientECH structure. The caller provides the ECH
-// version indicated by the client.
-func echUnmarshalClient(raw []byte) (*echClient, error) {
-	// Parse the payload as a ClientECH structure.
-	ech := new(echClient)
+// echUnmarshalClientOuter parses a ClientECH structure. The caller provides the
+// ECH version indicated by the client.
+func echUnmarshalClientOuter(raw []byte) (*echClientOuter, error) {
+	s := cryptobyte.String(raw)
+	ech := new(echClientOuter)
 	ech.raw = raw
 
+	// Make sure this is the outer handshake.
+	var variant uint8
+	if !s.ReadUint8(&variant) {
+		return nil, fmt.Errorf("error parsing ClientECH.type")
+	}
+	if variant != 0 {
```

Comment:
We seem to be using this enum a lot -- can we make it a constant so the code's more readable? E.g.,

```suggestion
	if variant != outerClientHelloVariant {
```

We could also then drop comments like the one on L376:

```
b.AddUint8(0) // Indicates ClientHelloOuter.
```

(This isn't blocking, of course. Feel free to resolve if there's no interest in this!)

Reply:
Done (good idea!).